rfcs/20210115-tfx-periodic-training.md

```diff
@@ -0,0 +1,445 @@
+
+*For example, assume a Span corresponds to a week of data, and a periodic
+pipeline run is scheduled weekly on Wednesday. The diagram below shows the
+timeline of the current week. The pipeline run on Wednesday picks up the version
+of last week’s data generated on Tuesday. Another version of that data arrives
+on Thursday, and on Friday a decision is made to refresh the model, so a
+backfill run is triggered manually which will materialize the latest version of
+last week’s data and retrain the current model.*
+
+<div style="text-align:center"><img src='20210115-tfx-periodic-training/timeline.png', width='600'></div>
+
+Backfilling a new Version of a Span that is out of rolling window range is
+unnecessary. After ExampleGen generates the latest Version of a Span within the
+rolling window, Resolver will retrieve the newest available (i.e., materialized)
+Version of each Span in the rolling window for Trainer.
+
+*For example, assume a Span corresponds to a day of data, and we have a Resolver
+that resolves the latest 3 days’ Span. On day 6, a model is generated based on
+data of day 4 to 6, and then an updated input data of day 5 is available. To
+backfill, we rerun the pipeline with specified Span id #5, ExampleGen will
+generate the new Version of Span 5, Resolver of that pipeline will retrieve
+ExampleGen’s output Spans of day 4 to 6 (not day 3 to 5, Resolver retrieves the
+latest window unless specified) for Trainer. Below shows the graph for backfill
+pipeline:*
+
+<div style="text-align:center"><img src='20210115-tfx-periodic-training/new_version.png', width='500'></div>
+
+Note that the normal run and backfill run need to have the same pipeline name
+for correct ML metadata tracking.
+
+##### Partial pipeline execution
+
+In certain backfill cases users may not want to run the entire pipeline. For
+example, when there are multiple spans in the rolling window that need to be
+backfilled, it’s not necessary to retrain the model multiple times. In such
+cases, user might want to (conceptually) divide the pipeline in two for a faster
+backfill:
+
+*   The first portion contains components like ExampleGen, StatisticsGen and
+    ExampleValidator. This part of the pipeline is scheduled for each Span.
+*   The second portion contains a Resolver plus components like Transform,
+    Trainer, Evaluator and Pusher. This part of the pipeline only needs to run
+    once for updating the model after all Spans are updated.
+
+We plan to provide the ability to specify a partial pipeline to execute at
+runtime (see Future Work below) to enable this workflow.
+
+An alternative approach is to define the two pipeline portions as TFX pipelines
+in their own right, to make backfill easier. This is not recommended because it
+makes the normal case of periodic training more brittle: the two pipelines must
+now be triggered in sequence by some non-TFX mechanism, thereby multiplying the
+possible fault conditions.
```

Comment:
How about using a Cross pipeline ResolverNode. Aka resolve from another pipeline. This allows you to do some kind of asyncronous support:
- ExampleGen pipeline runs every hour
- Training pipeline runs every day and takes latest 48 spans from the other pipeline.

Reply:
For async, it's kind of risky to break sync pipeline into multiple sub pipeline and schedule separately. Thus we only recommend break pipeline when there is [backfill needs](https://github.com/tensorflow/community/blob/875c04645f9029cb3c5d75bfdb8bf63e5560e9d9/rfcs/20210115-tfx-periodic-training.md#partial-pipeline-execution) , It has a lot of potential issues and not very robust. 
Internally we have a centralized orchestrator for async, if async is indeed needed in your case, stay tuned and we are working on it!

Reply:
Definetly interested on Async! Looking forward to that!

Reply:
@casassg I'm wondering what is the end goal behind the setup you suggest (compared to having the ExampleGen run every day with the training pipeline, and using a resolver to yield the latest 48 spans)?

Are you trying to reduce the end-to-end latency of your training pipeline by pre-materializing 23 of the 24 new spans? Or is there some other utility in doing the materialization incrementally (ex. load on whatever data source you're ingesting data from?)

Reply:
There's 2 reasons. 
1) data latency. By decoupling data processing from training, we can get faster latency for data needed (aka we just need to wait to generate 1 hour instead of having to generate a full day of data before starting our pipeline). This is the same principle and benefit explained in ASYNC pipelines, but using several pipelines and cross-pipeline dependencies to achieve it (this already is well supported within Airflow btw)

2) Better redundancy. Sometimes data can be delayed due to problems in data processing upstream. In this case, we may want to run the pipeline w the latest 24 spans (given its hour based, you are only 1 hour delayed vs unknown waiting time when issue there are several issues). This could be also achieved if we can trigger manually parts of the pipeline.

Reply:
Thanks - 1 makes sense and agreed that async pipelines is the more robust way of accomplishing this.

on the failure mode you describe in 2, is the "problems in data processing upstream" transient errors that upon retry will succeed? Because if not, and the data is simply not "ready"/accessible in the upstream source, the only benefit of more frequent ingestion is knowing about the upstream failure earlier and fixing it before the training run happens; so I don't quite follow the 'run the pipeline w the latest 24 spans' mitigation. 

(to provide broader context, I'm trying to validate the hypothesis that async will solve _all_ the use cases that you are tempted to solve with this cross-pipeline setup :)) 

Thanks for the discussion!

Reply:
For 2 what I meant is using Latest 24 available spans. This could be useful when refreshing your model is more time sensitive than it is having the latest data.

Example:
I want to deploy a model every day at 9PM  w 24 hours data (bc online metrics go down after 1 day due to changing data)
- 7PM - Data pipeline starts failing  - Alerts team
- 9PM - Model trains w data from DAY-1.7PM to DAY.7PM (latest 24 spans) - This avoids keeping an "expired" model around for too long and only uses 2 hour old data which is not terrible.
- 11PM - Fix data pipeline upstream and backfill data (ExampleGen) from 7PM to 11PM spans for next model.

I do think this can be achieved w Async pipelines (at least from what I understand) but just making sure.

Reply:
This can be done with sync pipeline too (but yep, with higher latency)

e.g., config ExampleGen pattern to grab a certain set of data, or allow Resolver to tolerate some missing data.

But agree, async can make it more automatic due to the centralized orchestration